File: 프로그래머스/lv3/43164. 여행경로/여행경로.py
A=[]
flag=0
def DFS(start, end, L, tickets, check,answer):
    global flag
    if L == len(tickets)-1:
        answer[L+1]=end
        A=answer
        if len(A)!=0:
            flag=1
            return

    for i in range(len(tickets)):
        if check[i] == 0 and end == tickets[i][0]:
            answer[L+1]=end
            check[i] = 1
            DFS(tickets[i][0], tickets[i][1], L + 1, tickets, check,answer)
            if flag==1:
                continue
            check[i]=0


def solution(tickets):
    global flag
    tickets.sort()
    answer = [""]*(len(tickets)+1)
    ans=[]
    check = [0] * len(tickets)
    for i in range(len(tickets)):
        flag=0
        st, end = tickets[i]
        if st == "ICN" and check[i] == 0:
            check[i] = 1
            answer[0]=st
            DFS(st, end, 0, tickets, check,answer)
            if flag==1:
                break
            check[i]=0
    return answer

File: 프로그래머스/lv3/43164. 여행경로/여행경로.java
import java.util.*;
class ticket implements Comparable<ticket>{
    String st;
    String end;
    public ticket(String st, String end){
        this.st=st;
        this.end=end;
    }
    
    @Override
    public int compareTo(ticket o){
        int x=(this.st).compareTo(o.st);
        if(x==0){
            x=(this.end).compareTo(o.end);
        }
        return x;
    }
    @Override
    public String toString(){
        return this.st+" "+this.end;
    }
}

class Solution {
    ArrayList<ticket> t=new ArrayList<>();
    String[] answer;
    int n=0;
    int[] ch;
    boolean flag=false;
    public String[] solution(String[][] tickets) {
        n=tickets.length;
        answer=new String[n+1];
        ch=new int[n];
        for(int i=0;i<tickets.length;i++){
            ticket tmp=new ticket(tickets[i][0],tickets[i][1]);
            t.add(tmp);
        }
        Collections.sort(t);
        for(int i=0;i<n;i++){
            System.out.println(t.get(i).toString());
        }
        int idx=0;
        for(int i=0;i<n;i++){
            ticket tmp=t.get(i);
            if((tmp.st).equals("ICN")){
                ch[i]=1;
                answer[0]=tmp.st;
                DFS(i,0);
                if(flag==true){
                    break;
                }
                ch[i]=0;
            }
        }
        return answer;
    }
    public void DFS(int nowIdx,int cnt){
        ticket now=t.get(nowIdx);
        String nowSt=now.st;
        String nowEnd=now.end;
        if(cnt==n-1){
            answer[cnt+1]=nowEnd;
            flag=true;
            return ;
        }
        
        for(int i=0;i<n;i++){
            ticket next=t.get(i);
            String nextSt=next.st;
            String nextEnd=next.end;
            
            if(ch[i]==0 && nowEnd.equals(nextSt)){
                answer[cnt+1]=nextSt;
                ch[i]=1;
                cnt+=1;
                DFS(i,cnt);
                if(flag){
                    return ;
                }
                cnt-=1;
                ch[i]=0;
            }
        }
        
    }
}